hummingbird/ml/operator_converters/_tree_implementations.py

```diff
@@ -67,13 +67,14 @@ class AbstractPyTorchTreeImpl(AbstracTreeImpl, torch.nn.Module):
     Abstract class definig the basic structure for tree-base models implemented in PyTorch.
     """
 
-    def __init__(self, tree_parameters, n_features, classes, n_classes):
+    def __init__(self, tree_parameters, n_features, classes, n_classes, is_anomaly_detection):
```

Comment:
Rather that do this, can you please add `anomaly_detection` to `BaseOperator`? We use that class for this purpose.

Reply:
Make sense! I totally missed the BaseOperator, indeed I can add that. But I wonder if I can remove those "anomaly_detection" arguments in those base implementations, because in `__init__` of `AbstractPyTorchTreeImpl` there is a condition on self.anomaly_detection, so if I remove all those arguments, it will always be False because it calls the "__init__" of BaseOperator first. What do you think?

Reply:
Yea I was checking the code and I don't think we can completely remove it, you are right. I was thinking that maybe we can move these code directly in the implementations in the `skl-iforest` file but I am not sure it will actually be better.

Reply:
I can either keep all those argument, or maybe do this: remove all those arguments, add a line `self.anomaly_detection = True` in each isolation forest implementation right before the `super().__init__()`, then in the BaseOperator, add a condition: `if not hasattr(self, 'anomaly_detection'): self.anomaly_detection = False`.

Which one do you think is better?

Reply:
Good idea! Do we actually need `if not hasattr(self, 'anomaly_detection'): self.anomaly_detection = False`? Since `BaseOperator` is called first, we can just have `anomaly_detection = False` by default and simply set it in the subclass, no?

Reply:
I don't think so, that's actually what I did at first then found out not working, there could be 2 situations: 1. if I set it True in the subclass **before** the `super().__init__()`, then it will be overwritten by `BaseOperator` and always False; 2. if I set it True in the subclass **after** the `super().__init__()`, then it can't change the condition in e.g. `AbstractPyTorchTreeImpl` because the `BaseOperator` is called first and had anomaly_detection = False. So in the above situation I need to add a condition to avoid being overwritten if it's already assigned value. What do you think?

Reply:
Ok I see your point now. What if we change the signature of `BaseOperator` and have it taking something like `regression=False, classification=False,anomaly_detection=False`. In this case we can directly set the behavior from the base class is known in advance before `__init__` or set if directly afterward.

Reply:
Indeed good idea to add the default parameters in `BaseOperator`, but don't understand how the last sentence can work. Now the class inheritance is like: `BaseOperator` => `AbstracTreeImpl` => `AbstractPyTorchTreeImpl` => `GEMMTreeImpl` => `GEMMIsolationForestImpl`. In `AbstractPyTorchTreeImpl` I need to do condition based on `anomaly_detection` which is default to be False in `BaseOperator` and I want to set it True in the last subclass `GEMMIsolationForestImpl`. I don't see how I can do that by simply adding the default parameters, maybe you mean something like `super().super().super()...__init__(anomaly_detection=True)`? But still it will be overwritten later when calling the `__init__`. And how about this: add default parameters in `BaseOperator` as you suggested, also add a `**kwargs` in parameters and __init__ calls in the base and every subclass in the chain, so that in the final subclass I can simply set that to True, and for other types of models it will be default to False if not specified. This should allow adding other parameters easier in the future as well, what do you think?

Reply:
Deal! :) I think this is a good mix between ease of implementation and generality.

Reply:
great! will push new commits later